### ISSUE TITLE
LeftCurly: Provide separate/unique token type for opening braces of anonymous inner class

### ISSUE THREAD

**Issue by eric-milles** [checkstyle/checkstyle#3324]
$ cat AnonInner.java
```java
import java.util.*;
class AnonInner
{ // no violation -- want opening brace of class to be on new line
    Comparator c = new Comparator()
    { // VIOLATION -- want the opening brace of anon. inner to be at end of previous line
        @Override public int compare(Object lhs, Object rhs)
        { // no violation -- want opening brace of method to be on new line
            return 0;
        }
    };
}
```
$ cat config.xml
```xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="LeftCurly">
            <message key="line.previous" value="Opening ''{0}'' should be at end of previous line" />
            <property name="option" value="eol" />
            <property name="tokens" value="LAMBDA,OBJBLOCK" />
        </module>
        <module name="LeftCurly">
            <message key="line.new" value="Opening ''{0}'' should be alone on a line" />
            <property name="option" value="nl" />
            <property name="tokens" value="CLASS_DEF,INTERFACE_DEF,ANNOTATION_DEF,ENUM_DEF,CTOR_DEF,METHOD_DEF,STATIC_INIT,
                                           LITERAL_IF,LITERAL_ELSE,LITERAL_FOR,LITERAL_DO,LITERAL_WHILE,LITERAL_SWITCH,LITERAL_SYNCHRONIZED,LITERAL_TRY,LITERAL_CATCH,LITERAL_FINALLY"
            />
        </module>
    </module>
</module>
```
$ java -jar checkstyle-7.0-all.jar -c config.xml AnonInner.java
Starting audit...
[ERROR] C:\Users\U0076777\AnonInner.java:3:1: Opening '{' should be at end of previous line [LeftCurly]
[ERROR] C:\Users\U0076777\AnonInner.java:5:5: Opening '{' should be at end of previous line [LeftCurly]
Audit done.
Checkstyle ends with 2 errors.

---------------

What I'm looking for is a way to isolate opening brace of anonymous inner class.  Currently, the only option available is OBJBLOCK.  However, this token type is also present under CLASS_DEF and a number of others that are already in use.

If I remove OBJBLOCK from my first LeftCurly definition, I get violations for opening braces on the same line for classes, interfaces, enums.  This is what we want.  However, I can't think of a way to generate a warning for the opening brace of Comparator anonymous inner class on the next line.  We want lambdas, array inits, and anon. inner classes to use the tighter style of brace at the end of the line.

AST for anon. inner is currently like this:
![checkstyle ast of anon inner class](https://cloud.githubusercontent.com/assets/18193802/16422934/be65b10e-3d1f-11e6-8ad0-f54d0e35cb8a.png)

**Comment by romani** [comment]
@eric-milles , I though about his for a while and I failed to make reasonable name for property to make it clear and make it work (processObjBlockOfInnnerBlockOnly - is weird and ..... ).

We do need new Check for this that specially validate RLcurly tokens. Example: http://checkstyle.sourceforge.net/config_whitespace.html#GenericWhitespace 

Please open new issue on this if you need such Check.

**Comment by eric-milles** [comment]
Would this issue be approved for work if we could agree on a config parameter name?  The token type route cannot work without a specific token and my understanding from other issues is that the AST is fixed.  That is, LeftCurly supports token type OBJBLOCK, but without a way to restrict checking to anon. inner classes or all other uses besides inners, the check will create false positives.

**Comment by eric-milles** [comment]
As far as the OBJBLOCK token goes, are there any situations besides a class initializer and an anon. inner class expression where this would appear?  I thought of static inits, but there is a STATIC_INIT token already.  So I'm trying to understand if OBJBLOCK for LeftCurly has any meaning besides anon. inner classes.

Or are you being careful of backwards compatibility because someone could have controlled the style of numerous tokens (CLASS_INIT, STATIC_INIT, etc.) though the OBJBLOCK token type?

**Comment by romani** [comment]
> Would this issue be approved for work if we could agree on a config parameter name?

lets discuss this in mail-list, if you find good way to name an property to make config looks good and reasonable - I will be ok.

> but without a way to restrict checking to anon. inner classes or all other uses besides inners, the check will create false positives.

yes

> inner class expression where this would appear?

I do not know all cases, wide testing will show that.

> So I'm trying to understand if OBJBLOCK for LeftCurly has any meaning besides anon. inner classes.


http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#OBJBLOCK

> Or are you being careful of backwards compatibility because someone could have controlled the style of numerous tokens (CLASS_INIT, STATIC_INIT, etc.) though the OBJBLOCK token type?

yes, we have to be compatible. We could brake compatibility only for good reason.

-------------

lets switch such discussions in mail-list it will hel

**Comment by eric-milles** [comment]
I tried to mail the devel and user lists and my message was bounced by both.


My goal is to be able to enforce the LeftCurly check on anonymous inner classes -- separately from other constructs.  In the AST, OBJBLOCK is the token that appears under EXPR and LITERAL_NEW for an anon. inner class.  There is a screenshot of this in the issue.  The problem with the current check is that OBJBLOCK appears in multiple situations.  So, without additional config, we get violations on all sorts of other block types (if, for, class def, etc.).

It was proposed to add a Boolean config option that would restrict OBJBLOCK checks to when they appear for anonymous inner classes.  This was not implemented as the option name got very cumbersome.

I would propose an option that is a token list.  Maybe "objblockParents" or "objblockTokens".  It could default to "CLASS_DEF,INTERFACE_DEF,ANNOTATION_DEF,..." which would result in todays check behavior.  Then, I could set it to be "LITERAL_NEW", which would only ever have an OBJBLOCK child for an anonymous inner class instance.

----------

Concrete example:
```xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="LeftCurly">
            <message key="line.previous" value="Opening ''{0}'' should be at end of previous line" />
            <property name="option" value="eol" />
            <property name="tokens" value="LAMBDA,OBJBLOCK" />
            <property name="objblockTokens" value="LITERAL_NEW" />
        </module>
        <module name="LeftCurly">
            <message key="line.new" value="Opening ''{0}'' should be alone on a line" />
            <property name="option" value="nl" />
            <property name="tokens" value="CLASS_DEF,INTERFACE_DEF,ANNOTATION_DEF,ENUM_DEF,CTOR_DEF,METHOD_DEF,STATIC_INIT,
                                           LITERAL_IF,LITERAL_ELSE,LITERAL_FOR,LITERAL_DO,LITERAL_WHILE,LITERAL_SWITCH,LITERAL_SYNCHRONIZED,LITERAL_TRY,LITERAL_CATCH,LITERAL_FINALLY"
            />
        </module>
    </module>
</module>
```
This config would check { at the end of the line for lambdas and anon. inners only.  For all other constructs, we expect to see { alone on a new line so that you can match the closing } in the same column of the editor.  Also, this creates simpler line-wrapping rules.

**Comment by romani** [comment]
No, I do not like that design.  It will over-complicate this Check, we already have few cases in other Checks where desire to address all cases resulted in bad and unclear design, users unsatisfied also. 

We need new Check for that validation. New Checks for now should pass testing in sevntu-checkstyle project.

**Comment by eric-milles** [comment]
Wow, that was unexpected.  You proposed the boolean flag, not me.  Now you are telling me a new check is required to handle one case of a left curly that is already checked as part of this check.  The only thing that is lacking in LeftCurlyCheck is the ability to isolate an instance of OBJBLOCK.

You flag and my token list option both support today's check behavior by default.  All that is left is the supposed complication of the check.  You haven't even seen a patch or pull request for the change, so how can you be so sure the check will be complicated beyond understanding?

**Comment by romani** [comment]
>  I though about his for a while and I failed to make reasonable name for property to make it clear and make it work (processObjBlockOfInnnerBlockOnly - is weird and ..... ).

I failed to make it in good way, so I proposed to to look at new Check creation.

>  You haven't even seen a patch or pull request for the change

patch does not matter is design is not good. I do not think we could make smth good with  new option.

> so how can you be so sure the check will be complicated beyond understanding?

I just read config. Config should be clear even without reading documentation. And I have some experience in support of this project so I know what users struggle with.
Unfortunately there is nobody in project to whom we can ask to second opinion.
.... but lets try ....
@rnveach , can you share your opinion on proposed design?

**Comment by rnveach** [comment]
@romani Sorry for the delay, almost missed my name on this.

The base of the discussion seems to arise from wanting to differentiate between Method, Class, and Anonymous Class curlies and allow the users to specify them independently. This stems from the problem that the AST doesn't directly support this which is the user's main form of customization in the config.

> We could brake compatibility only for good reason.

Correct me if I'm wrong, I believe we said we are breaking all kinds of backwards compatibility in CS version 8? Maybe making each of these a new and separate token would be possible then. The downside is this would complicate our code some, but would allow the user even greater customization without the need for a million new properties on each check. We could separate out other tokens that are needed.

I do agree with romani that only adding a new property might not be the way to go. It is either only going to target a small subset of the tokens, or going to have to be customizable enough that it some how applies to all the tokens.

Another idea that seems more realistic is we could do something similar to what we did to the naming checks. Everything in our AST for the class/variable names are `IDENT`, so we overcame targeting a specific case by making almost 11 checks. We could split `LeftCurly` (and its brother `RightCurly`) into `ClassLeftCurly`, `MethodLeftCurly`, etc. The majority of unique tokens left over could stay in another check, either `LeftCurly` or some newly named check, like tokens `IF`, `ELSE`, etc. This would make the check clear, not need additional properties, and allow the user to target specific cases.

&lt;/rambling&gt;

**Comment by romani** [comment]
> Correct me if I'm wrong, I believe we said we are breaking all kinds of backwards compatibility in CS version 8?

no , we will change API in that version so all extensions and plugins will be affected. We did braking changes for Checks in minor version changes. It is ok if that is very good reason.

> Maybe making each of these a new and separate token would be possible then

I doubt, we already have few ANTLR issues and everyone afraid to touch. Just think .... to fix one Check you are going to change AST tree and damage whole universe of Checks. It is not fare. Especially if that known that requested design could be easily done as separate Check.

> We could split 

We can not remove Check that now, it is used in whole world and works fine for them.
We can create new Checks to work more precise - I am ok. That is what I am suggesting - create new Check for required needs. Some functional overlap with existing Check is completely ok.

**Comment by guillerodriguez** [comment]
Old issue but still relevant in current version. 

There were some comments of possible fixes involving:

1. Modifying the existing LeftCurly check so that one can (optionally) specify a set of “parent tokens”. 
2. Modifying the AST so that we can target anonymous inner classes using a more specific token than OBJBLOCK. 
3. Adding a new Check

Approach 2 would break backwards compatibility, which is not desired. I think 3 would provide a solution for this specific issue, but it would not solve other similar cases. On the other hand 1 would fully preserve compatibility, AST is not changed, existing checks would not break, and at the same time would be generic enough to fix this and possibly other similar issues. The new “parentTokens” property would be optional — if not used the default behaviour would be the same as of today. 

Is there an option for this to be reconsidered? I would be happy to work in the implementation if that helps.
